### ISSUE TITLE
How to remove successfully loaded files from list?

### ISSUE THREAD

**Issue by Hellek** [pqina/vue-filepond#12]
Hi! Can you explain how i can do this? I tried to do this for last 4 hours, but it was effortless

1) I can't get name and ID for each loaded file. processfile() just fire that file is loaded (even if file get error for oversize), but i cant get information which of multiple.
2) error() not get information about file which was errored
3) Also i cant find how i can remove file from list (not from server)

**Comment by rikschennink** [comment]
You can use the "v-on:addfile" or "v-on:error" events to access data on the added or errored file, see the [callbacks list](https://pqina.nl/filepond/docs/patterns/api/filepond-instance/#callbacks) to see which parameters are passed.

```html
<file-pond v-on:addfile="handleFilePondAddFile"></file-pond>
```

```js
new Vue({
  el: '#app',
  methods: {
    handleFilePondAddFile: function(error, file) {
      console.log('add file', error, file);
    }
  },
  components: {
    FilePond
  }
});
```

To remove files, Make a `ref` to the pond instance, the example below will remove a file after a file has been added.

```html
<file-pond ref="pond" v-on:addfile="handleFilePondAddFile"></file-pond>
```

```js
new Vue({
  el: '#app',
  methods: {
    handleFilePondAddFile: function(error, file) {
      console.log('add file', error, file);
      setTimeout(() => {
        console.log('remove file');
        this.$refs.pond.removeFile()
      }, 500);
    }
  },
  components: {
    FilePond
  }
});
```

**Comment by Hellek** [comment]
Arrrrgh, by bad. I define callback func with parentheses
`<file-pond @addfile="handleFilePondAddFile()"></file-pond>`
and method no gets arguments. Your template example works. Tnx
`<file-pond @addfile="handleFilePondAddFile"></file-pond>`

**Comment by Hellek** [comment]
I found out that this variant works not ax expected.

Error reproduce:
1. Making an 10 files upload.
2. When first files are uploaded some of last files stoping with error.
3. Cause `this.$refs.pond.removeFile()` executes for all files in the list - not just for each successfully loaded file

My code
```
<template>
	<FilePond
		...
		@processfile="handleFileUpload"
		...
	/>
</template>

<script>
export default {
	methods: {
		...
		handleFileUpload(error, file) {
			if (error === null) {
				this.$refs.pond.removeFile()
			}
		}
		...
	}
}
</script>
```

**Comment by rikschennink** [comment]
@Hellek `removeFile` without parameters removes the first file in the files list, [see docs](https://pqina.nl/filepond/docs/patterns/api/filepond-instance/#removing-files). Better to pass the id of the file.

```js
handleFileUpload(error, file) {
    
    if (error) {
        return;
    }

    this.$refs.pond.removeFile(file.id);
}
```

**Comment by Hellek** [comment]
Yep, that's it. I'm looking for tomething like ID. My bad, don't found it at docs. Thanks!

**Comment by Hellek** [comment]
It's solve problem, but partly. Before delete file from filepond i add filename to another list (of loaded files).
```
this.files.push(file.filename)
this.$refs.pond.removeFile(file.id);
```
And sometime it still produce an error
```
filepond.js?66d1:6868 Uncaught TypeError: Cannot read property 'filename' of null
    at Object.itemProcessed [as DID_COMPLETE_ITEM_PROCESSING] (filepond.js?66d1:6868)
    at eval (filepond.js?66d1:1536)
    at Array.forEach (<anonymous>)
    at eval (filepond.js?66d1:1535)
    at eval (filepond.js?66d1:1278)
    at Array.forEach (<anonymous>)
    at Object._write (filepond.js?66d1:1277)
    at eval (filepond.js?66d1:1305)
    at Array.forEach (<anonymous>)
    at Object._write (filepond.js?66d1:1303)
```

this.files.push(file.filename) goes before deletion, and that's strange...
I suppose I can solve it with this.$nextTick() - but it's not right way

**Comment by rikschennink** [comment]
Is it possible that an error is thrown and the file item object is not there? Does it throw the same error without the remove File statement?

**Comment by Hellek** [comment]
1) It make a test - error ocures certainly here.
2) There is no error in this two cases:
```
handleFileUpload(error, file) {
    if (error) { return; }
    ...code goes here...
}
```
2.1)
`this.files.push(file.filename)`
2.2)
```
this.files.push(file.filename)
this.$nextTick(() => {
    this.$refs.pond.removeFile(file.id);
})
```
3) There is error in this cases
3.1) Just for one file (if i load alot of them)
```
this.files.push(file.filename)
this.$refs.pond.removeFile(file.id);
```
3.2) For few files
```
this.files.push(file.filename)
this.$refs.pond.removeFile();
```

**Comment by rikschennink** [comment]
It's probably because the `DID_COMPLETE_ITEM_PROCESSING` flow hasn't finished yet. By removing in the next tick it ensures FilePond is done with the file. I'll see if I can make these events asynchronous so you can be certain that file FilePond is done.

**Comment by Hellek** [comment]
How it can be unfinished if "handleFileUpload" callback on @processfile.
If i understand docs right, @processfile calls when item processing is complete.

Anyway, i don't understand how it can be, no matter "this.$refs.pond.removeFile(file.id)" sync or async, it calls after "this.files.push(file.filename)".

I csn just guess, that object which comes to callback function "handleFileUpload(error, file)" already removed but filepond call func for it again

**Comment by Hellek** [comment]
@rikschennink yep, you are right. This error throws inside filepond code. I removed my part `this.files.push(file.filename)` where i trying to get "filename" property, and error not vanished.

Is filepond has an event, which mean "file upload is completely done"? Cause as we found out @processfile not mean that all actions are finished.

**Comment by rikschennink** [comment]
I'll see if I can fix this by throwing the `processfile` event a tiny bit later

**Comment by rikschennink** [comment]
Just published version `1.7.3`, FilePond events are now async, in my tests, this seems to solve the issue.

**Comment by Hellek** [comment]
@rikschennink I load 1.7.3 manually tried to reproduce that error. For now it isn't. Looks that it works fine without this.nextTick(). Hoping that this not get back again 😃
